### PR TITLE
Introducing `danger pr` command

### PR DESCRIPTION
**Pull request by JuanitoFatas**
This Pull Requests adds a `danger pr` command that works for Pull Requests. So that in the future we can rebuild `danger local` to make it only work for local git repositories.

Can try it out if you clone this branch and:

```
Work with Open PR:

$ bin/danger pr --use-pr=https://github.com/danger/danger/pull/615

Work with Merged PR:

$ bin/danger pr --use-pr=https://github.com/danger/danger/pull/613

Works with forks, too:

$ bin/danger pr --use-pr=https://github.com/JuanitoFatas/danger/pull/3
```

--

<details>
<summary>Implementation details</summary>
The implementation of `danger pr` command is copied from `danger local` command. And by introducing `LOCAL_GIT_PR_URL`, `LocalGitRepo` will rely on this ENV to dispatch what to do and kept `danger local` remains unchanged.
</details>

<details>
<summary>Regarding to Rubocop:</summary>

These 2 rules I don't agree.

```
Offenses:

lib/danger/ci_source/support/pull_request_finder.rb:84:11: C: Align select with begin on line 82.
          select { |pr| pr.state == "open" }.
          ^^^^^^
lib/danger/ci_source/support/pull_request_finder.rb:85:11: C: Align map with begin on line 82.
          map do |open_pr|
          ^^^
```

Now

```ruby
@open_pull_requests ||= begin
  client.pull_requests(repo_slug).
    select { |pr| pr.state == "open" }.
    map do |open_pr|
      OpenPullRequest.new(
        open_pr.number.to_s,
        open_pr.title,
        open_pr.head.sha,
        open_pr.base.sha
      )
    end
end
```
Rubocop

```ruby
@open_pull_requests ||= begin
  client.pull_requests(repo_slug).
                        select { |pr| pr.state == "open" }.
                        map do |open_pr|
                          OpenPullRequest.new(
                            open_pr.number.to_s,
                            open_pr.title,
                            open_pr.head.sha,
                            open_pr.base.sha
                          )
                        end
end
```

What do you prefer? I can stick to what Rubocop wants if everyone feels rubocop's better.

=> Changed to new style in 419732c.
</details>

### CODE REVIEW

**Comment by orta** [comment]
Now's style

**Comment by orta** [comment]
Cool - yeah, this looks good to me 👍

**Comment by orta** [comment]
and not even a breaking change ;)

File: spec/lib/danger/ci_sources/local_git_repo_spec.rb [hunk 1]
```diff
@@ -55,103 +57,140 @@ def run_in_repo(merge_pr: true, squash_and_merge_pr: false)
...
-      it "gets the repo address when it uses git@" do
+      it "gets the repo slug when it uses https" do
+        run_in_repo do
+          `git remote add origin https://github.com/danger/danger.git`
+
+          result = source(valid_env)
+
+          expect(result.repo_slug).to eq("danger/danger")
+        end
+      end
```
Comment (JuanitoFatas) [inline spec/lib/danger/ci_sources/local_git_repo_spec.rb]: Yea, this newly added test catch a bug 🎉 , will fix.

**Comment by orta** [comment]
WRT: remote finding, you could switch to ignore all of that and expect people to put a full PR URL in there? This makes it really easy to test fork PRs?

**Comment by JuanitoFatas** [comment]
> WRT: remote finding, you could switch to ignore all of that and expect people to put a full PR URL in there? This makes it really easy to test fork PRs?

👍 Yea, that's better. 

However, the failure is actually telling us if someone set their remote by HTTPS instead of SSH, the repo slug we found is not correct. Only failed on Circle, wonder if Circle did something weird.

**Comment by JuanitoFatas** [comment]
> Only failed on Circle, wonder if Circle did something weird.

Circle mutated our remote url 😱 I made the Regexp more robust to also handle it in 35c0783 <3

**Comment by JuanitoFatas** [comment]
- [ ] danger pr works for fork PRs

**Comment by JuanitoFatas** [comment]
- [ ] fail fase if cannot find repo slug

**Comment by orta** [comment]
That looks _awesome_ 👍

**Comment by orta** [comment]
Should fix https://github.com/danger/danger/issues/445 https://github.com/danger/danger/issues/289 https://github.com/danger/danger/issues/370 https://github.com/danger/danger/issues/276

File: lib/danger/ci_source/support/pull_request_finder.rb [hunk 2]
```diff
@@ -1,25 +1,88 @@
...
+          LocalPullRequest.new(most_recent_squash_and_merged_pull_request)
+        elsif remote && remote_pull_request
+          RemotePullRequest.new(
+            remote_pull_request.number.to_s,
+            remote_pull_request.head.sha,
+            remote_pull_request.base.sha
+          )
+        else
+          NoPullRequest.new
+        end
```
Comment (orta) [inline lib/danger/ci_source/support/pull_request_finder.rb]: elegant

File: lib/danger/commands/systems.rb [hunk 3]
```diff
@@ -16,10 +16,12 @@ def run
...
-      ci_sources = registry.all(:class)
-                           .select { |klass| klass.inheritance_tree.map(&:name).include? :CI }
-                           .reject { |source| source.name == :CI }
-                           .reject { |source| source.name == :LocalGitRepo }
+      ci_sources = begin
+        registry.all(:class)
+          .select { |klass| klass.inheritance_tree.map(&:name).include? :CI }
+          .reject { |source| source.name == :CI }
+          .reject { |source| source.name == :LocalGitRepo }
+      end
```
Comment (orta) [inline lib/danger/commands/systems.rb]: this reads much nicer

File: spec/lib/danger/commands/pr_spec.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      argv = CLAide::ARGV.new(["--use-pr=https://github.com/danger/danger/pull/615"])
+
+      result = described_class.new(argv)
+
+      expect(result).to have_instance_variables(
+        "@pr_url" => "https://github.com/danger/danger/pull/615"
+      )
+    end
+  end
+end
```
Comment (orta) [inline spec/lib/danger/commands/pr_spec.rb]: nice work - I've always found testing the commands pretty tricky - think this is a good answer 👍

Comment (JuanitoFatas) [inline spec/lib/danger/commands/pr_spec.rb]: TBH this only tests the states of the PR command class, not how it runs 😓 Still haven't figured out how to test how it runs 🏃

File: spec/lib/danger/ci_sources/support/pull_request_finder_spec.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  def two_kinds_of_merge_log
+    IO.read("spec/fixtures/ci_source/support/two-kinds-of-merge-both-present.log")
+  end
+
+  def open_pull_requests_info
+    require "ostruct"
+    JSON.parse(
+      IO.read("spec/fixtures/ci_source/support/danger_danger_pr_518.json"),
+      object_class: OpenStruct
```
Comment (JuanitoFatas) [inline spec/lib/danger/ci_sources/support/pull_request_finder_spec.rb]: This is a TIL for me, the hash can use like objects.

http://stackoverflow.com/a/32470168

**Comment by JuanitoFatas** [comment]
LGTM 
🛳⛵🚢⛴🛥🚤

**Comment by orta** [comment]
Yep, given it a third run through! Let's do it - @JuanitoFatas want to do a version bump + release?

**Pull request merged**